pkg/clients/edgeconnect/client.go

```diff
@@ -0,0 +1,278 @@
+	}
+
+	transport.Proxy = func(req *http.Request) (*url.URL, error) {
+		return proxyUrl, nil
+	}
+	httpClient := http.Client{Transport: transport}
+	ctx := context.WithValue(context.Background(), oauth2.HTTPClient, httpClient)
+
+	return NewClient(clientID, clientSecret, WithContext(ctx))
+}
+
+func WithBaseURL(url string) func(*client) {
+	return func(c *client) {
+		c.baseURL = url
+	}
+}
+
+func WithTokenURL(url string) func(*client) {
+	return func(c *client) {
+		c.TokenURL = url
+	}
+}
+
+// WithContext can set context for client
+// NB: via context you can override default http client to add Proxy or CA certificates
+func WithContext(ctx context.Context) func(*client) {
+	return func(c *client) {
+		c.ctx = ctx
+	}
+}
+
+// ServerError represents an error returned from the server (e.g. authentication failure).
+type ServerError struct {
+	Code    int          `json:"code,omitempty"`
+	Message string       `json:"message,omitempty"`
+	Details DetailsError `json:"details"`
+}
+
+// DetailsError represents details of errors
+type DetailsError struct {
+	ConstraintViolations ConstraintViolations `json:"constraintViolations"`
+	MissingScopes        []string             `json:"missingScopes,omitempty"`
+}
+
+// ConstraintViolations represents constraint violation errors
+type ConstraintViolations struct {
+	Message           string `json:"message"`
+	Path              string `json:"path,omitempty"`
+	ParameterLocation string `json:"parameterLocation,omitempty"`
+}
+
+// Error formats the server error code and message.
+func (e ServerError) Error() string {
+	if len(e.Message) == 0 && e.Code == 0 {
+		return "unknown server error"
+	}
+
+	return fmt.Sprintf("edgeconnect server error %d: %s", int64(e.Code), e.Message)
+}
+
+type serverErrorResponse struct {
+	ErrorMessage ServerError `json:"error"`
+}
+
+func CloseBodyAfterRequest(response *http.Response) {
+	if response != nil && response.Body != nil {
+		err := response.Body.Close()
+		if err != nil {
+			return
+		}
+	}
+}
+
+func (c *client) handleErrorResponseFromAPI(response []byte, statusCode int) error {
+	se := serverErrorResponse{}
+	if err := json.Unmarshal(response, &se); err != nil {
+		return errors.WithMessagef(err, "response error, can't unmarshal json response %d", statusCode)
+	}
+
+	return se.ErrorMessage
+}
+
+func (c *client) getServerResponseData(response *http.Response) ([]byte, error) {
+	responseData, err := io.ReadAll(response.Body)
+	if err != nil {
+		return nil, errors.WithMessage(err, "error reading response")
+	}
+
+	if response.StatusCode != http.StatusOK &&
+		response.StatusCode != http.StatusCreated {
+		return responseData, c.handleErrorResponseFromAPI(responseData, response.StatusCode)
+	}
+
+	return responseData, nil
+}
+
+// GetEdgeConnect returns edge connect if it exists
+func (c *client) GetEdgeConnect(edgeConnectId string) (GetResponse, error) {
+	url := c.getEdgeConnectUrl(edgeConnectId)
+
+	resp, err := c.Client(c.ctx).Get(url)
```

Comment:
Is it guaranteed that `Client()` always returns a working client? Should there be a nil check?